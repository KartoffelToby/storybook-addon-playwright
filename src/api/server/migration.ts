import path from 'path';
import glob from 'fast-glob';
import { readFileSync } from 'jsonfile';
import {
  ActionSet,
  ScreenshotOptions,
  PlaywrightDataStories,
  ScreenshotSetting,
  PlaywrightData,
} from '../../typings';
import { setStoryOptions } from './services/utils';
import { nanoid } from 'nanoid';
import { getStoryPlaywrightFileInfo, saveStoryFile } from './utils';

interface ScreenshotDataV0 extends ScreenshotSetting {
  title: string;
  hash?: string;
  index?: number;
  options: ScreenshotOptions & { cursor?: boolean };
  props?: {
    name: string;
    value?: never;
  }[];
}

interface V0StoryData {
  actionSets?: ActionSet & { description?: string }[];
  screenshots?: ScreenshotDataV0[];
  hash: string;
}

interface V0 {
  [story: string]: V0StoryData;
}

const migrateToV1 = (data: V0, version: string) => {
  const storyIds = Object.keys(data);

  const newData: PlaywrightData = {
    version: version,
  };

  newData.stories = storyIds.reduce((newStoryData, storyId) => {
    const storyData = data[storyId] as V0StoryData;

    if (!Object.keys(storyData).length) {
      return newStoryData;
    }

    newStoryData[storyId] = {};

    if (storyData.actionSets && storyData.actionSets.length) {
      newStoryData[storyId].actionSets = storyData.actionSets.map(
        (actionSet) => {
          (actionSet as ActionSet).title = actionSet.description;
          delete actionSet.description;
          return actionSet as ActionSet;
        },
      );
    }

    if (storyData.screenshots) {
      newStoryData[storyId].screenshots = storyData.screenshots.map(
        (screenshot) => {
          const props = screenshot.props
            ? screenshot.props.reduce((obj, prop) => {
                obj[prop.name] = prop.value;
                return obj;
              }, {})
            : undefined;

          if (screenshot.options) {
            const cursor = screenshot.options.cursor;
            delete screenshot.options.cursor;
            screenshot.screenshotOptionsId = setStoryOptions(
              newData,
              'screenshotOptions',
              screenshot.options,
            );
            if (cursor) {
              screenshot.browserOptionsId = setStoryOptions(
                newData,
                'browserOptions',
                { cursor: true },
              );
            }
            delete screenshot.options;
          }
          delete screenshot.hash;
          return {
            ...screenshot,
            id: nanoid(15),
            props,
          };
        },
      );
    }

    return newStoryData;
  }, {} as PlaywrightDataStories);

  return newData;
};

export const migrateFile = (file: string, version: string) => {
  const data = readFileSync(file);
  if (!data.version) {
    console.log(`\nMigrating ${path.parse(file).name} to v${version}...`);
    return migrateToV1(data, version);
  }
  return undefined;
};

export const migration = () => {
  const packagePath = path.resolve(__dirname, '../../../package.json');

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const version = (require(packagePath).version as string)
    .split('.')[0]
    .toString();

  console.log('\nMigrating to v' + version + ' ...');

  const files = glob.sync(['**/*.playwright.json', '!node_modules/**']);

  files.forEach((file) => {
    const fileInfo = getStoryPlaywrightFileInfo(file);
    const data = migrateFile(file, version);
    if (data) {
      saveStoryFile(fileInfo, data);
    }
  });

  console.log(`\nEnd of migration.`);
};
