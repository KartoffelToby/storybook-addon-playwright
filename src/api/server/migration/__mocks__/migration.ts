export const migration = jest.fn();
