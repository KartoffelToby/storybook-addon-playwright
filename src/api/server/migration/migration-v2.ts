import {
  PlaywrightData,
  ActionSet,
  ScreenshotData,
  StoryAction,
} from '../../../typings';
import { nanoid } from 'nanoid';

interface V1ScreenshotData extends ScreenshotData {
  actions: StoryAction[];
}

export interface PlaywrightStoryData {
  actionSets?: ActionSet[];
  screenshots?: V1ScreenshotData[];
}

export type PlaywrightDataStories = { [id: string]: PlaywrightStoryData };

interface V1PlaywrightData extends PlaywrightData {
  stories?: PlaywrightDataStories;
}

export const migrationV2 = (data: V1PlaywrightData, version: string) => {
  data.version = version;
  if (data.stories) {
    for (const storyKey in data.stories) {
      const story = data.stories[storyKey];

      if (story.actionSets) {
        story.actionSets.forEach((actionSet) => {
          actionSet.actions.forEach((action) => {
            delete action.id;
          });
        });
      }

      if (story.screenshots) {
        story.screenshots.forEach((screenshot) => {
          if (screenshot.actions) {
            screenshot.actionSets = [
              {
                actions: screenshot.actions.map((action) => {
                  delete action.id;
                  return action;
                }),
                id: nanoid(12),
                title: `${screenshot.title} actions`,
              },
            ];
            delete screenshot.actions;
          }
        });
      }
    }
  }

  return data;
};
