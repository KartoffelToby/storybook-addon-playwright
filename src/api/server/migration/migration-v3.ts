import { PlaywrightData } from '../../../typings';

export const migrationV3 = (data: PlaywrightData, version: string) => {
  data.version = version;
  return data;
};
