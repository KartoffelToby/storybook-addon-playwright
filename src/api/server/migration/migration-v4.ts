import { PlaywrightData } from '../../../typings';

export const migrationV4 = (data: PlaywrightData, version: string) => {
  data.version = version;
  return data;
};
