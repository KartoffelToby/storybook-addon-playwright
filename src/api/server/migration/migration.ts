import path from 'path';
import glob from 'fast-glob';
import { readFileSync } from 'jsonfile';
import { getStoryPlaywrightFileInfo, saveStoryFile } from '../utils';
import { migrateToV1 } from './migration-v1';
import { migrationV2 } from './migration-v2';
import { getVersion } from '../utils';

export const migrateFile = (file: string, version: string) => {
  let data = readFileSync(file);

  console.log(`\nMigrating ${path.parse(file).name} to v${version}...`);

  if (!data.version) {
    data = migrateToV1(data, '1');
  }

  if (data.version === '1') {
    data = migrationV2(data, '2');
  }

  return data;
};

export const migration = () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const version = getVersion();

  console.log('\nMigrating to v' + version + ' ...');

  const files = glob.sync(['**/*.playwright.json', '!node_modules/**']);

  files.forEach((file) => {
    const fileInfo = getStoryPlaywrightFileInfo(file);
    const data = migrateFile(file, version);
    if (data) {
      saveStoryFile(fileInfo, data);
    }
  });

  console.log(`\nEnd of migration.`);
};
