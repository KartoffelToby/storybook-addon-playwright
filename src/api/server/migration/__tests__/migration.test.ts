import { readFileSync } from 'jsonfile';
import { migrateToV1 } from '../migration-v1';
import { migrationV2 } from '../migration-v2';
import path from 'path';
import { migrateFile } from '../migration';

describe('migration', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should migrate from v0 to v1', () => {
    const v0FilePath = path.resolve(__dirname, './migration-v0-data.json');
    const data = readFileSync(v0FilePath);
    const newData = migrateToV1(data, '1');
    expect(newData).toMatchSnapshot();
  });

  it('should migrate from v1 to v2', () => {
    const v1FilePath = path.resolve(__dirname, './migration-v1-data.json');
    const data = readFileSync(v1FilePath);
    const newData = migrationV2(data, '2');
    expect(newData).toMatchSnapshot();
  });

  it('should migrate from v0 to v2', () => {
    const v0FilePath = path.resolve(__dirname, './migration-v0-data.json');
    const newData = migrateFile(v0FilePath, '2');
    expect(newData).toMatchSnapshot();
  });
});
