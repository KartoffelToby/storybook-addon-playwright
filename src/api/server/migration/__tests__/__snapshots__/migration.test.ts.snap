// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migration should migrate from v0 to v1 1`] = `
Object {
  "browserOptions": Object {
    "id-2": Object {
      "cursor": true,
    },
  },
  "screenshotOptions": Object {
    "id-1": Object {
      "clip": Object {
        "height": 600,
        "width": 600,
        "x": 0,
        "y": 0,
      },
    },
  },
  "stories": Object {
    "with-actionSets": Object {
      "actionSets": Array [
        Object {
          "actions": Array [
            Object {
              "args": Object {
                "timeout": 500,
              },
              "id": "4ps_NuM6B2",
              "name": "waitForTimeout",
            },
          ],
          "id": "k71GTTcObZen",
          "title": "delay 500",
        },
      ],
    },
    "with-defaults": Object {
      "screenshots": Array [
        Object {
          "actions": Array [
            Object {
              "args": Object {
                "options": Object {
                  "button": "middle",
                },
              },
              "id": "yryt9qdYHQ",
              "name": "click",
            },
          ],
          "browserOptionsId": "id-2",
          "browserType": "chromium",
          "id": "id-3",
          "index": 4,
          "props": Object {
            "color": "red",
            "size": 20,
          },
          "screenshotOptionsId": "id-1",
          "title": "should apply styles",
        },
      ],
    },
  },
  "version": "1",
}
`;

exports[`migration should migrate from v0 to v2 1`] = `
Object {
  "browserOptions": Object {
    "id-6": Object {
      "cursor": true,
    },
  },
  "screenshotOptions": Object {
    "id-5": Object {
      "clip": Object {
        "height": 600,
        "width": 600,
        "x": 0,
        "y": 0,
      },
    },
  },
  "stories": Object {
    "with-actionSets": Object {
      "actionSets": Array [
        Object {
          "actions": Array [
            Object {
              "args": Object {
                "timeout": 500,
              },
              "name": "waitForTimeout",
            },
          ],
          "id": "k71GTTcObZen",
          "title": "delay 500",
        },
      ],
    },
    "with-defaults": Object {
      "screenshots": Array [
        Object {
          "actionSets": Array [
            Object {
              "actions": Array [
                Object {
                  "args": Object {
                    "options": Object {
                      "button": "middle",
                    },
                  },
                  "name": "click",
                },
              ],
              "id": "id-8",
              "title": "should apply styles actions",
            },
          ],
          "browserOptionsId": "id-6",
          "browserType": "chromium",
          "id": "id-7",
          "index": 4,
          "props": Object {
            "color": "red",
            "size": 20,
          },
          "screenshotOptionsId": "id-5",
          "title": "should apply styles",
        },
      ],
    },
  },
  "version": "2",
}
`;

exports[`migration should migrate from v1 to v2 1`] = `
Object {
  "browserOptions": Object {
    "c99BcUTVviggy9H9vRFM8": Object {
      "cursor": true,
    },
  },
  "screenshotOptions": Object {
    "jTfiAoRWN1vGRuthlPSRl": Object {
      "clip": Object {
        "height": 600,
        "width": 600,
        "x": 0,
        "y": 0,
      },
    },
  },
  "stories": Object {
    "with-defaults": Object {
      "actionSets": Array [
        Object {
          "actions": Array [
            Object {
              "args": Object {
                "x": 497,
                "y": 113,
              },
              "name": "mouse.move",
            },
          ],
          "id": "OSzX1CUEn71R",
          "title": "vertical scroll down",
        },
      ],
      "screenshots": Array [
        Object {
          "actionSets": Array [
            Object {
              "actions": Array [
                Object {
                  "args": Object {
                    "selector": "html",
                  },
                  "name": "click",
                },
              ],
              "id": "id-4",
              "title": "should show on click actions",
            },
          ],
          "browserOptionsId": "c99BcUTVviggy9H9vRFM8",
          "browserType": "chromium",
          "id": "DFNYJhhPYuQHTDX",
          "index": 1,
          "props": Object {
            "showOnCreate": false,
          },
          "screenshotOptionsId": "jTfiAoRWN1vGRuthlPSRl",
          "title": "should show on click",
        },
      ],
    },
  },
  "version": "2",
}
`;
